internal/config/endpoints.go

```diff
@@ -25,176 +35,143 @@ type Endpoint struct {
+		return err
+	}
+	defer r.Body.Close()
+	return json.NewDecoder(r.Body).Decode(target)
+
+}
+
+func GetEndpoints() map[string]Endpoint {
+	//TODO detect when running in Cloud Run and GenerateConfigFromAPI()
+	if len(allEndpoints) == 0 {
+		allEndpoints = GenerateConfigFromEndpoints()
+	}
+	return allEndpoints
+}
+
+func GenerateConfigFromEndpoints() map[string]Endpoint {
+	//Used by CLI to pull endpoint configs from Cloud Run enpoints.
+	e := make(map[string]Endpoint)
+	ie := new(map[string]innerEndpoint)
+	getJson("https://global.gcping.com/api/endpoints", &ie)
+	for _, s := range *ie {
+		e[s.Region] = Endpoint(s)
+	}
+	return e
+}
+
+func GenerateConfigFromAPI() map[string]Endpoint {
+	//Used by Cloud Run Endpoints to pull endpoint configs from Cloud Run Admin API
+	log.Print("Using Cloud Run Admin API to generate Endpoints config.")
+	ctx := context.Background()
+	runService, err := run.NewService(ctx)
+	// TODO: Get project name from Cloud Run metadata service if not defined in env variable
+	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
+
+	if err != nil {
+		panic(err)
+	}
+	// List Services
+	resp, err := runService.Namespaces.Services.List("namespaces/" + projectID).Fields("items(status/address/url,metadata(labels,name),spec(template/metadata/annotations))").LabelSelector("env=prod").Do()
+
+	if err != nil {
+		panic(err)
+	}
+
+	s, _ := json.MarshalIndent(resp.Items, "", "\t")
+
+	var endpoints []Endpoint
+
+	json.Unmarshal([]byte(s), &endpoints)
+
+	var EndpointsMap = make(map[string]Endpoint)
+
+	// Add global endpoint to map if env is defined.
+	globalUrl := os.Getenv("GLOBAL_ENDPOINT")
+	if globalUrl != "" {
+		var Global Endpoint
+		Global.Region = "global"
+		Global.RegionName = "Global External HTTPS Load Balancer"
+		Global.URL = os.Getenv("GLOBAL_ENDPOINT")
+		EndpointsMap[Global.Region] = Global
+	}
+
+	for _, s := range endpoints {
+		EndpointsMap[s.Region] = s
+	}
+
+	return EndpointsMap
+}
+
+func (es *Endpoint) UnmarshalJSON(data []byte) error {
+	//Custom unmarshal function for Endpoint data
+	type labelsInner struct {
+		Location string `json:"cloud.googleapis.com/location"`
+	}
+
+	type annotationsInner struct {
+		RegionName string `json:"region-name"`
+	}
+
+	type templateMetadataInner struct {
+		Annotations annotationsInner `json:"annotations`
+	}
+
+	type templateInner struct {
+		Metadata templateMetadataInner `json:"metadata"`
+	}
+
+	type specInner struct {
+		Template templateInner `json:"template"`
+	}
+
+	type metadataInner struct {
+		Labels labelsInner `json:"labels"`
+		Name   string      `json:"name"`
+	}
+
+	type addressInner struct {
+		Url string `json:"url"`
+	}
+
+	type statusInner struct {
+		Address addressInner `json:"address"`
+	}
+
+	type nestedService struct {
+		Metadata metadataInner `json:"metadata"`
+		Status   statusInner   `json:"status"`
+		Spec     specInner     `json:"spec"`
+	}
+
+	var ns nestedService
+
+	if err := json.Unmarshal(data, &ns); err != nil {
+		return err
+	}
```

Comment:
If I understand this correctly, the `nestedService` is a superset of the `Endpoint` which we are unmarshaling and so we are defining a custom unmarshal for `Endpoint` with inline types for `nestedService`

From a readability standpoint I would rather define those structs outside of the function (with the appropriate `json:` annotations you have) and then in line 98 unmarshal to `[]nestedService`, finally make a function to loop over the []nestedService and call a funciton like `nestedServiceToEnpoint` or `endpointFromNestedService` to translate one in to the other. There is some duplication in memory usage, but it is clearer and allows for both structs to evolve as need be.

The way it is currently written the structs are very tightly coupled, and the inline type definitions make reading difficult (imo)

Reply:
I revisited this portion with the following changes:
- renamed `nestedService` to `nestedEndpoint` for better readability/association with  the un-nested `Endpoint`
- moved the struct definitions for the nested JSON to a new file `nested_endpoints.go`
- removed the custom unmarshal function for `nestedEndpoint`. unmarshaling to a `nestedEndpoint` is handled by the default unmarshall func, following the `nestedEndpoint` and associated structs. 
- Added an un-nesting func `unNestEndpoint` func.